File: aoc24/d7/solu.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Dir struct {
	name    string
	parent  *Dir
	child   []*Dir
	content [][]string
}

func NewDir(name string, parent *Dir) *Dir {
	return &Dir{
		name:    name,
		parent:  parent,
		child:   []*Dir{},
		content: [][]string{},
	}
}

func (d *Dir) Sz() int {
	res := 0
	for _, c := range d.content {
		if len(c) > 0 {
			value, err := strconv.Atoi(c[0])
			if err == nil {
				res += value
			}
		}
	}
	for _, c := range d.child {
		res += c.Sz()
	}
	return res
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	input := []string{}
	for scanner.Scan() {
		input = append(input, scanner.Text())
	}

	if len(input) > 0 {
		input = input[1:]
	}

	testInput := input
	cdir := NewDir("/", nil)
	c1dir := cdir

	res := []int{}

	for len(testInput) > 0 {
		line := testInput[0]
		testInput = testInput[1:]
		parts := strings.Split(line, " ")
		if parts[0] == "$" {
			if parts[1] == "cd" {
				if parts[2] == ".." {
					c1dir = c1dir.parent
				} else {
					for _, c := range c1dir.child {
						if c.name == parts[2] {
							c1dir = c
							break
						}
					}
				}
			}
			continue
		}

		tp := parts[0]
		name := parts[1]
		if tp == "dir" {
			c1dir.child = append(c1dir.child, NewDir(name, c1dir))
		} else {
			c1dir.content = append(c1dir.content, []string{tp, name})
		}
	}

	NEEDED_SPACE := 30000000
	USED_SPACE := 70000000 - cdir.Sz()

	var rec func(d *Dir)
	rec = func(d *Dir) {
		sz := d.Sz()
		// if sz <= 100000 { // solu1
		//     res = append(res, sz)
		// }
		if USED_SPACE+sz >= NEEDED_SPACE { // solu2
			res = append(res, sz)
		}
		if len(d.child) > 0 {
			for _, c := range d.child {
				rec(c)
			}
		}
	}

	rec(cdir)

	fmt.Println(res)
	minVal := res[0]
	for _, val := range res {
		if val < minVal {
			minVal = val
		}
	}
	fmt.Println(minVal)
}

File: aoc24/d7/solu.py
X = """190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20""".splitlines()

X = open("input.txt","r").read().splitlines()

def rec(x, n=5, res=[]) -> list:
    if len(x)==n:
        res.append(x)
        return res
    rec(x+"*", n=n, res=res)
    rec(x+"+", n=n, res=res)
    rec(x+"|", n=n, res=res)
    return res

res=[]
for c,x in enumerate(X):
    print(c,"/",len(X), end="\r")
    ans, nums = x.split(": ")
    nums = nums.split(" ")
    combs = rec("",n=len(nums)-1, res=[])
    evals = nums[0]
    for c in combs:
        evals = nums[0]
        for n,o in zip(nums[1:],c):
            if o == "|":
                evals = str(evals) + n
            else:
                evals = eval(str(evals)+o+n)

        if eval(str(evals)) == int(ans):
            res.append(int(ans))
            break

print(sum(res))

